File: formal/coqimp/task05/extraction_hs/Fappli_IEEE_extra.hs
module Fappli_IEEE_extra where

import qualified Prelude
import qualified BinInt
import qualified BinPos
import qualified Datatypes
import qualified Fappli_IEEE
import qualified Fcore_Zaux

coq_BofZ :: Prelude.Integer -> Prelude.Integer -> Prelude.Integer -> Fappli_IEEE.Coq_binary_float
coq_BofZ prec emax n =
  Fappli_IEEE.binary_normalize prec emax Fappli_IEEE.Coq_mode_NE n 0 Prelude.False

coq_ZofB :: Prelude.Integer -> Prelude.Integer -> Fappli_IEEE.Coq_binary_float -> Prelude.Maybe
            Prelude.Integer
coq_ZofB _ _ f =
  case f of {
   Fappli_IEEE.B754_zero _ -> Prelude.Just 0;
   Fappli_IEEE.B754_finite s m e0 ->
    (\fO fP fN n -> if n Prelude.== 0 then fO () else
                   if n Prelude.> 0 then fP n else
                   fN (Prelude.negate n))
      (\_ -> Prelude.Just (Fcore_Zaux.cond_Zopp s ((\x -> x) m)))
      (\e -> Prelude.Just
      ((Prelude.*) (Fcore_Zaux.cond_Zopp s ((\x -> x) m))
        (BinInt._Z__pow_pos (Fcore_Zaux.radix_val Fcore_Zaux.radix2) e)))
      (\e -> Prelude.Just
      (Fcore_Zaux.cond_Zopp s
        (BinInt._Z__div ((\x -> x) m)
          (BinInt._Z__pow_pos (Fcore_Zaux.radix_val Fcore_Zaux.radix2) e))))
      e0;
   _ -> Prelude.Nothing}

coq_ZofB_range :: Prelude.Integer -> Prelude.Integer -> Fappli_IEEE.Coq_binary_float ->
                  Prelude.Integer -> Prelude.Integer -> Prelude.Maybe Prelude.Integer
coq_ZofB_range prec emax f zmin zmax =
  case coq_ZofB prec emax f of {
   Prelude.Just z ->
    case (Prelude.&&) (BinInt._Z__leb zmin z) (BinInt._Z__leb z zmax) of {
     Prelude.True -> Prelude.Just z;
     Prelude.False -> Prelude.Nothing};
   Prelude.Nothing -> Prelude.Nothing}

pos_pow :: Prelude.Integer -> Prelude.Integer -> Prelude.Integer
pos_pow x y =
  (\fI fO fH n -> if n Prelude.== 1 then fH () else
                   if Prelude.odd n
                   then fI (n `Prelude.div` 2)
                   else fO (n `Prelude.div` 2))
    (\y0 -> BinPos._Pos__mul x (BinPos._Pos__square (pos_pow x y0)))
    (\y0 -> BinPos._Pos__square (pos_pow x y0))
    (\_ -> x)
    y

coq_Bparse :: Prelude.Integer -> Prelude.Integer -> Prelude.Integer -> Prelude.Integer ->
              Prelude.Integer -> Fappli_IEEE.Coq_binary_float
coq_Bparse prec emax base m e =
  let {emin = (Prelude.-) ((Prelude.-) ((\x -> x) ((\x -> 2 Prelude.* x Prelude.+ 1) 1)) emax) prec}
  in
  (\fO fP fN n -> if n Prelude.== 0 then fO () else
                   if n Prelude.> 0 then fP n else
                   fN (Prelude.negate n))
    (\_ -> coq_BofZ prec emax ((\x -> x) m))
    (\p ->
    case BinInt._Z__ltb ((Prelude.*) e (BinInt._Z__log2 ((\x -> x) base))) emax of {
     Prelude.True -> coq_BofZ prec emax ((Prelude.*) ((\x -> x) m) ((\x -> x) (pos_pow base p)));
     Prelude.False -> Fappli_IEEE.B754_infinity Prelude.False})
    (\p ->
    case BinInt._Z__ltb
           ((Prelude.+) ((Prelude.*) e (BinInt._Z__log2 ((\x -> x) base)))
             (BinInt._Z__log2_up ((\x -> x) m))) emin of {
     Prelude.True -> Fappli_IEEE.B754_zero Prelude.False;
     Prelude.False ->
      Fappli_IEEE.coq_FF2B prec emax
        (case Fappli_IEEE.coq_Fdiv_core_binary prec ((\x -> x) m) 0 ((\x -> x) (pos_pow base p)) 0 of {
          (,) p0 lz ->
           case p0 of {
            (,) mz ez ->
             (\fO fP fN n -> if n Prelude.== 0 then fO () else
                   if n Prelude.> 0 then fP n else
                   fN (Prelude.negate n))
               (\_ -> Fappli_IEEE.F754_nan Prelude.False 1)
               (\mz0 ->
               Fappli_IEEE.binary_round_aux prec emax Fappli_IEEE.Coq_mode_NE
                 (Datatypes.xorb Prelude.False Prelude.False) mz0 ez lz)
               (\_ -> Fappli_IEEE.F754_nan Prelude.False 1)
               mz}})})
    e

File: formal/coqimp/task06/extraction_ml/Fappli_IEEE_extra.ml
open BinInt
open BinNums
open BinPos
open Datatypes
open Fappli_IEEE
open Fcore_Zaux

(** val coq_BofZ : coq_Z -> coq_Z -> coq_Z -> binary_float **)

let coq_BofZ prec emax n =
  binary_normalize prec emax Coq_mode_NE n Z0 false

(** val coq_ZofB : coq_Z -> coq_Z -> binary_float -> coq_Z option **)

let coq_ZofB _ _ = function
| B754_zero _ -> Some Z0
| B754_finite (s, m, e0) ->
  (match e0 with
   | Z0 -> Some (cond_Zopp s (Zpos m))
   | Zpos e ->
     Some (Z.mul (cond_Zopp s (Zpos m)) (Z.pow_pos (radix_val radix2) e))
   | Zneg e ->
     Some (cond_Zopp s (Z.div (Zpos m) (Z.pow_pos (radix_val radix2) e))))
| _ -> None

(** val coq_ZofB_range :
    coq_Z -> coq_Z -> binary_float -> coq_Z -> coq_Z -> coq_Z option **)

let coq_ZofB_range prec emax f zmin zmax =
  match coq_ZofB prec emax f with
  | Some z -> if (&&) (Z.leb zmin z) (Z.leb z zmax) then Some z else None
  | None -> None

(** val pos_pow : positive -> positive -> positive **)

let rec pos_pow x = function
| Coq_xI y0 -> Pos.mul x (Pos.square (pos_pow x y0))
| Coq_xO y0 -> Pos.square (pos_pow x y0)
| Coq_xH -> x

(** val coq_Bparse :
    coq_Z -> coq_Z -> positive -> positive -> coq_Z -> binary_float **)

let coq_Bparse prec emax base m e =
  let emin = Z.sub (Z.sub (Zpos (Coq_xI Coq_xH)) emax) prec in
  (match e with
   | Z0 -> coq_BofZ prec emax (Zpos m)
   | Zpos p ->
     if Z.ltb (Z.mul e (Z.log2 (Zpos base))) emax
     then coq_BofZ prec emax (Z.mul (Zpos m) (Zpos (pos_pow base p)))
     else B754_infinity false
   | Zneg p ->
     if Z.ltb (Z.add (Z.mul e (Z.log2 (Zpos base))) (Z.log2_up (Zpos m))) emin
     then B754_zero false
     else coq_FF2B prec emax
            (let (p0, lz) =
               coq_Fdiv_core_binary prec (Zpos m) Z0 (Zpos (pos_pow base p))
                 Z0
             in
             let (mz, ez) = p0 in
             (match mz with
              | Zpos mz0 ->
                binary_round_aux prec emax Coq_mode_NE (xorb false false) mz0
                  ez lz
              | _ -> F754_nan (false, Coq_xH))))
